postprocessing/processors/job_handling.py

```diff
@@ -72,3 +106,56 @@ def determine_success_local(configuration, out_err):
             data["error"] = f"REDUCTION: {error_line}"
 
     return success, data
+
+
+def get_memory_limit_mb(configuration):
+    """
+    Get the memory limit in Megabytes based on a percentage of the available system memory
+    @param Configuration configuration: configuration
+    @return float: memory limit in MB
+    """
+    mem_total = psutil.virtual_memory().total
+    mem_percentage = configuration.system_mem_limit_perc
+    return mem_total * mem_percentage * CONVERSION_FACTOR_BYTES_TO_MB
+
+
+def get_total_memory_usage(proc):
+    """
+    Get the total memory usage in bytes of process ``proc`` and its children
+    @param Popen proc: process
+    @return float: memory usage in bytes
+    """
+    # Start with the memory usage of the parent process
+    total_memory = proc.memory_info().rss
+    # Iterate through all child processes and add their memory usage
+    for child in proc.children(recursive=True):
+        total_memory += child.memory_info().rss
+    return total_memory
+
+
+def terminate_or_kill_process_tree(pid, timeout=3):
```

Comment:
This function makes sense and I like how it escalates terminating the processes. The question I have is how the logging will look in the end. If there are no subprocesses it should be very clear in the logs, but when there are children it may be confusing to figure out which is the parent based on the logs. 

Question: Should terminate/kill processes log at the error level or warning level? I don't know if warning will show up as an error in webmon.

Reply:
Let me look into what the logs look like and make clear that distinction between parent and child processes if needed.

We can discuss whether warning or error is the right level here. I was thinking that warning is more appropriate if we see this as standard mode of operation now that post-processing agent will terminate/kill memory intensive subprocesses.

The part that makes it show up as an error for the run in WebMon is where we add a message to the run reduction _error_ log file, see the functions `local_submission` and `determine_success_local`.